src/ServantElm.hs

```diff
@@ -0,0 +1,215 @@
+import GHC.Desugar (AnnotationWrapper)
+import GHC.ExecutionStack (Location (functionName))
+import GHC.Generics (Generic, to)
+import Lens.Micro ((^.))
+import Network.Wai ()
+import Network.Wai.Handler.Warp (run)
+import Servant (Application, Get, JSON, Proxy (Proxy), Server, serve)
+import Servant.API
+import Servant.Foreign
+  ( GenerateList,
+    HasForeign (Foreign),
+    HasForeignType (..),
+    PathSegment (unPathSegment),
+    Req,
+    Segment (unSegment),
+    SegmentType (Cap, Static),
+    camelCase,
+    listFromAPI,
+    path,
+    reqFuncName,
+    reqMethod,
+    reqReturnType,
+    reqUrl,
+  )
+import Text.PrettyPrint.Leijen.Text
+  ( Doc,
+    comma,
+    dquotes,
+    encloseSep,
+    equals,
+    hsep,
+    indent,
+    lbrace,
+    lbracket,
+    line,
+    lparen,
+    parens,
+    pretty,
+    punctuate,
+    rbrace,
+    rbracket,
+    rparen,
+    space,
+    text,
+    textStrict,
+    vsep,
+    (<$>),
+    (<+>),
+  )
+import Prelude hiding ((<$>))
+
+data LangElm
+
+instance Elm a => HasForeignType LangElm ElmDefinition a where
+  typeFor _ _ proxyA = toElmDefinition proxyA
+
+elmForAPI ::
+  ( HasForeign LangElm ElmDefinition api,
+    GenerateList
+      ElmDefinition
+      (Foreign ElmDefinition api)
+  ) =>
+  Proxy api ->
+  [Doc]
+elmForAPI api =
+  map endpointInfoToElmQuery $
+    listFromAPI (Proxy :: Proxy LangElm) (Proxy :: Proxy ElmDefinition) api
+
+elmRecord :: [Doc] -> Doc
+elmRecord = encloseSep (lbrace <> space) (line <> rbrace) (comma <> space)
+
+indent4Spaces :: Doc -> Doc
+indent4Spaces = indent 4
+
+endpointInfoToElmQuery :: Req ElmDefinition -> Doc
+endpointInfoToElmQuery requestInfo =
+  funcDef
+  where
+    funcDef =
+      vsep
+        [ fnName <+> ":" <+> typeSignature,
+          fnName <+> args <+> equals,
+          indent4Spaces elmRequest
+        ]
+
+    fnName = textStrict . camelCase $ requestInfo ^. reqFuncName
+
+    typeSignature =
+      mkTypeSignature requestInfo
+
+    args =
+      mkArgs requestInfo
+
+    elmRequest =
+      mkRequest requestInfo
+
+elmList :: [Doc] -> Doc
+elmList [] = lbracket <> rbracket
+elmList ds = lbracket <+> hsep (punctuate (line <> comma) ds) <$> rbracket
+
+stext :: Text -> Doc
+stext = text . fromStrict
+
+mkUrl :: [Segment ElmDefinition] -> Doc
+mkUrl segments =
+  urlBuilder
+    <$> (indent4Spaces . elmList)
+      (map segmentToDoc segments)
+  where
+    urlBuilder :: Doc
+    urlBuilder = "Url.Builder.crossOrigin urlBase" :: Doc
+
+    segmentToDoc :: Segment ElmDefinition -> Doc
+    segmentToDoc s =
+      case unSegment s of
+        Static path ->
+          dquotes (stext (unPathSegment path))
+        Cap arg ->
+          error
+            "TODO implement-2" -- for captures, not needed now
+
+elmTypeRefToDoc :: TypeRef -> Doc
+elmTypeRefToDoc = \case
+  RefPrim elmPrim -> elmPrimToDoc elmPrim
+  RefCustom (TypeName typeName) -> pretty typeName
+
+elmTypeParenDoc :: TypeRef -> Doc -- does it make sense to be a fn
```

Comment:
Why not? It's used more than once, so it seems fine to have this.

Reply:
done